{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "_CW-SGYgGXyh"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Подготовка данных\n"
      ],
      "metadata": {
        "id": "_CW-SGYgGXyh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://raw.githubusercontent.com/knapweedss/hse21_hw1/main/data/scaffolds.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4JcUjvnbGcWK",
        "outputId": "9097473c-66e2-4dcc-c581-232201b5424e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-19 10:57:25--  https://raw.githubusercontent.com/knapweedss/hse21_hw1/main/data/scaffolds.fasta\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3967630 (3.8M) [text/plain]\n",
            "Saving to: ‘scaffolds.fasta.1’\n",
            "\n",
            "scaffolds.fasta.1   100%[===================>]   3.78M  --.-KB/s    in 0.04s   \n",
            "\n",
            "2021-12-19 10:57:25 (86.8 MB/s) - ‘scaffolds.fasta.1’ saved [3967630/3967630]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_gGEkv/gms2_linux_64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XXhmjrRSGhtP",
        "outputId": "caaf69b0-9f18-4673-ff38-04aef9e61f07"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-19 10:57:28--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_gGEkv/gms2_linux_64.tar.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11261783 (11M) [application/x-gzip]\n",
            "Saving to: ‘gms2_linux_64.tar.gz.1’\n",
            "\n",
            "gms2_linux_64.tar.g 100%[===================>]  10.74M  3.14MB/s    in 4.0s    \n",
            "\n",
            "2021-12-19 10:57:32 (2.70 MB/s) - ‘gms2_linux_64.tar.gz.1’ saved [11261783/11261783]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_gGEkv/gm_key_64.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Lc9bIblOmh4s",
        "outputId": "b06d4ae9-086f-4ac6-fc36-eae5951a005e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-19 10:57:35--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_gGEkv/gm_key_64.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 84 [application/x-gzip]\n",
            "Saving to: ‘gm_key_64.gz’\n",
            "\n",
            "\rgm_key_64.gz          0%[                    ]       0  --.-KB/s               \rgm_key_64.gz        100%[===================>]      84  --.-KB/s    in 0s      \n",
            "\n",
            "2021-12-19 10:57:35 (11.3 MB/s) - ‘gm_key_64.gz’ saved [84/84]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d gm_key_64.gz"
      ],
      "metadata": {
        "id": "jJ1u6E3OmlDw",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "58a35422-5bca-410f-f74f-e11718d41e70"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gzip: gm_key_64 already exists; do you wish to overwrite (y or n)? y\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzvf gms2_linux_64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-mHBA-kRmrCc",
        "outputId": "adc4dbde-9d6b-47d9-cf0a-1291441219f6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gms2_linux_64/\n",
            "gms2_linux_64/gms2.pl\n",
            "gms2_linux_64/README_GMS2.txt\n",
            "gms2_linux_64/gmhmmp2_key\n",
            "gms2_linux_64/mgm_4.mod\n",
            "gms2_linux_64/gmhmmp2\n",
            "gms2_linux_64/INSTALL\n",
            "gms2_linux_64/mgm_11.mod\n",
            "gms2_linux_64/biogem\n",
            "gms2_linux_64/LICENSE\n",
            "gms2_linux_64/mgm_25.mod\n",
            "gms2_linux_64/compp\n",
            "gms2_linux_64/mgm_15.mod\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cp  -v  gm_key_64   ~/.gmhmmp2_key"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hHAMEeA1mtMT",
        "outputId": "ac3be79c-6488-4d91-e2af-b07ed966e788"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "'gm_key_64' -> '/root/.gmhmmp2_key'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Установка E-utilities (для скачивания последовательностей из NCBI)"
      ],
      "metadata": {
        "id": "hsjQhypOm48m"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sh -c \"$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KjGnVQ9Xm6KR",
        "outputId": "4227cf50-1c82-4dff-fc24-7d4764b0d1c8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## ## Скачиваем данные по близкородственной бактерии T.oleivorans"
      ],
      "metadata": {
        "id": "VYwk5HesNgLw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gb  >  T_oleivorans_MIL_1.gbk\n",
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gene_fasta  >  T_oleivorans_MIL_1.genes.fasta\n",
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format fasta_cds_aa  >  T_oleivorans_MIL_1.proteins.fasta"
      ],
      "metadata": {
        "id": "3lBs0hkVNj2S"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем BLAST и устанавливаем пакет seqkit"
      ],
      "metadata": {
        "id": "VaKq3RrhnXSl"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt update\n",
        "!apt install ncbi-blast+\n",
        "\n",
        "!wget https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
        "!tar -xzvf seqkit_linux_amd64.tar.gz\n",
        "!chmod a+x seqkit"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uSxUyKTXm-Au",
        "outputId": "4ea9258f-9c57-4780-d72d-615adff67956"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33m\r0% [Working]\u001b[0m\r            \rHit:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease\n",
            "\u001b[33m\r0% [Connecting to archive.ubuntu.com] [Connecting to security.ubuntu.com (91.18\u001b[0m\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Connecting to archive.ubuntu.com] [Connecting to\u001b[0m\r                                                                               \rIgn:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n",
            "\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Connecting to archive.ubuntu.com] [Connecting to\u001b[0m\r                                                                               \rIgn:3 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n",
            "\u001b[33m\r0% [1 InRelease gpgv 3,626 B] [Connecting to archive.ubuntu.com] [Connecting to\u001b[0m\r                                                                               \rHit:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release\n",
            "Hit:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n",
            "Hit:6 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease\n",
            "Hit:7 http://security.ubuntu.com/ubuntu bionic-security InRelease\n",
            "Hit:8 http://archive.ubuntu.com/ubuntu bionic InRelease\n",
            "Hit:10 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n",
            "Hit:11 http://archive.ubuntu.com/ubuntu bionic-updates InRelease\n",
            "Hit:13 http://archive.ubuntu.com/ubuntu bionic-backports InRelease\n",
            "Hit:14 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease\n",
            "Hit:15 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "82 packages can be upgraded. Run 'apt list --upgradable' to see them.\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "ncbi-blast+ is already the newest version (2.6.0-1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 82 not upgraded.\n",
            "--2021-12-19 10:59:05--  https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 192.30.255.113\n",
            "Connecting to github.com (github.com)|192.30.255.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211219T105905Z&X-Amz-Expires=300&X-Amz-Signature=e4896c6e1c07d9a1d9a2a675ac821f0b09a572d935c964f554cf28761e539e05&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-12-19 10:59:05--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211219T105905Z&X-Amz-Expires=300&X-Amz-Signature=e4896c6e1c07d9a1d9a2a675ac821f0b09a572d935c964f554cf28761e539e05&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6650850 (6.3M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz.1’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   6.34M  --.-KB/s    in 0.09s   \n",
            "\n",
            "2021-12-19 10:59:06 (72.0 MB/s) - ‘seqkit_linux_amd64.tar.gz.1’ saved [6650850/6650850]\n",
            "\n",
            "seqkit\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Скачиваем БД белков SwissProt"
      ],
      "metadata": {
        "id": "TnSvkM6BnkyQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
        "!gzip -d uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qpBvh6zNnex-",
        "outputId": "a4d940b1-ddbc-41a9-b9be-405e3b22c568"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-19 10:59:10--  https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
            "Resolving ftp.uniprot.org (ftp.uniprot.org)... 128.175.240.195\n",
            "Connecting to ftp.uniprot.org (ftp.uniprot.org)|128.175.240.195|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 90527596 (86M) [application/x-gzip]\n",
            "Saving to: ‘uniprot_sprot.fasta.gz’\n",
            "\n",
            "uniprot_sprot.fasta 100%[===================>]  86.33M  36.5MB/s    in 2.4s    \n",
            "\n",
            "2021-12-19 10:59:13 (36.5 MB/s) - ‘uniprot_sprot.fasta.gz’ saved [90527596/90527596]\n",
            "\n",
            "gzip: uniprot_sprot.fasta already exists; do you wish to overwrite (y or n)? y\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Обработка данных"
      ],
      "metadata": {
        "id": "M7tRwloCL3Uf"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Предсказать расположение белок-кодирующих генов в геноме нашей бактерии"
      ],
      "metadata": {
        "id": "GNwyhPQVTPr3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./gms2_linux_64/gms2.pl  --seq scaffolds.fasta  --genome-type bacteria  --fnn genes.fasta  --faa proteins.fasta"
      ],
      "metadata": {
        "id": "-Xy3TSC3M35l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### получение файлов .fasta с генами и белками"
      ],
      "metadata": {
        "id": "dcVA-oXLL903"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in T_oleivorans_MIL_1.proteins.fasta  -out T_oleivorans_MIL_1.proteins"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NgWFD2DUnqt4",
        "outputId": "7bdc2242-c2c5-46c4-fb97-eecddbc1886f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/19/2021 11:01:41\n",
            "New DB name:   /content/T_oleivorans_MIL_1.proteins\n",
            "New DB title:  T_oleivorans_MIL_1.proteins.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 3666 sequences in 0.1052 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Определяем функции белков нашей бактерии через сравнение с белками из T.oleivorans MIL-1"
      ],
      "metadata": {
        "id": "zXQ59X1NMKb1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!blastp  -query proteins.fasta  -db T_oleivorans_MIL_1.proteins  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "pfJ4F1soMGRV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Определяем функции белков нашей бактерии через поиск по БД SwissProt"
      ],
      "metadata": {
        "id": "XKH5pq8WOPxr"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Смотрим, какие белки нашей бактерии (из Баренцева моря) не имели похожих белков из бактерии T.oleivorans MIL-1\n",
        "* Для этого получаем список уникальных номеров генов бактерии из Барнцева моря, коротые имели хотя бы один похожий белок бактерии из Сицилли"
      ],
      "metadata": {
        "id": "d4INYhGhOkCz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_MIL_1.txt | sort -n | uniq > proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "2rzaAj0FOQZu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./seqkit grep  --invert-match  -f proteins.with_hits_from_MIL_1.txt  proteins.fasta   -o proteins.without_MIL_1.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EXFph2FSOnEL",
        "outputId": "8b207549-62b5-40cb-f05b-930685014391"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[INFO]\u001b[0m 3337 patterns loaded from file\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Посмотрим на статистику"
      ],
      "metadata": {
        "id": "c5ldQD_DOr7m"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "- Общее количество белков в нашей бактерии:"
      ],
      "metadata": {
        "id": "QhWYjxvuTZZM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yP_w2rQYOy9U",
        "outputId": "3d82df99-56c5-4b46-bebb-8fcdfa385465"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3620\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "- Количество белков, которые имели схожие белки из бактерии MIL-1"
      ],
      "metadata": {
        "id": "iWeala_KPCk9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nl7_LqdqO3qF",
        "outputId": "06d8f75f-ba3d-4dff-80f2-3b4fd65ad100"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3337 proteins.with_hits_from_MIL_1.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "- Без аннотации функции"
      ],
      "metadata": {
        "id": "nHjNfkiLOO-p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.without_MIL_1.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Kldw5eT_O9Cd",
        "outputId": "8910ed1f-3d26-409d-9ee6-903903d4a5fb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "283\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Сравнить все белки (протеом) нашей бактерии с аннотированными белками из БД SwissProt\n",
        "* Индексируем белки SwissProt:"
      ],
      "metadata": {
        "id": "eitg1SIsTqUz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in uniprot_sprot.fasta  -out uniprot_sprot"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "L5qn2kaJQCjr",
        "outputId": "1b05d332-d6b5-4e17-e591-f12420e735b2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/19/2021 11:06:07\n",
            "New DB name:   /content/uniprot_sprot\n",
            "New DB title:  uniprot_sprot.fasta\n",
            "Sequence type: Protein\n",
            "Deleted existing Protein BLAST database named /content/uniprot_sprot\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 565928 sequences in 13.4377 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Определеям какие белки из нашей бактерии похожи на белки из БД SwissProt "
      ],
      "metadata": {
        "id": "Jtn6Thr7Tu5O"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!time  blastp  -query proteins.without_MIL_1.fasta  -db uniprot_sprot  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_SwissProt.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "acfm8IJvQHGC",
        "outputId": "af8ee851-56ca-47da-b79e-287a958ed301"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "real\t9m26.077s\n",
            "user\t9m24.990s\n",
            "sys\t0m0.365s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Посмотрим, сколько белков удалось аннотировать через поиск по SwissProt:"
      ],
      "metadata": {
        "id": "hPzVbMulSYzR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_SwissProt.txt | sort | uniq | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gK1NbG6MSZZf",
        "outputId": "97b3ead2-5d14-422f-c2c4-d8b5c8561180"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "52\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### файлы, которые нужно залить на гитхаб \n",
        "* gms2.lst  --  координаты всех предсказанных генов\n",
        "* proteins.fasta  --  аминокислотные последовательности всех предсказанных генов\n",
        "* scaffolds.hits_from_MIL_1.txt  --  информация о схожести белков из нашей бактерии с белками из бактерии MIL-1\n",
        "* scaffolds.hits_from_SwissProt.txt  -- информация о схожести белков из нашей бактерии с белками из БД SwissProt"
      ],
      "metadata": {
        "id": "HBt4QV5NShfb"
      }
    }
  ]
}